File: 2024/01/aoc24_1_2.jl
# AoC 2024 - Day 1 - Task 2
# https://erik-skopp.de/AdventofCode/2024/1/
# https://github.com/eskopp/AdventOfCode/blob/main/2024/01/aoc24_1_2.jl
#
# Day 1 - Historian Hysteria
# result: 20373490

# Change the working directory to the specified path
cd("2024/01")

# Arrays for unique and target values
unique_values = String[]
target_values = String[]

# Read the file and split data into columns
open("input.in", "r") do file
    for line in eachline(file)
        columns = split(strip(line), "   ") # Split columns based on triple spaces
        if length(columns) == 2
            push!(unique_values, columns[1])
            push!(target_values, columns[2])
        end
    end
end

# Count how often values from `unique_values` appear in `target_values`
frequency_map = Dict(value => count(==(value), target_values) for value in unique_values)

# Create a list of weighted counts
weighted_counts = Int[]
for (value, count) in frequency_map
    if count != 0
        push!(weighted_counts, parse(Int, value) * count)
    end
end

# Calculate the total sum of weighted counts
total_sum = sum(weighted_counts)

# Output the result
println(total_sum)

File: 2024/01/aoc24_1_2.py
# AoC 2024 - Day 1 - Task 2
# https://erik-skopp.de/AdventofCode/2024/1/
# https://github.com/eskopp/AdventOfCode/blob/main/2024/01/aoc24_1_2.py
#
# Day 1 - Historian Hysteria
# result: 20373490

import os

# Set the working directory
os.chdir(os.path.join("2024", "01"))

# Lists for unique and target values
unique_values = []
target_values = []

# Read the file and split data into columns
with open("input.in", "r") as file:
    for line in file:
        line = line.strip().split("   ")
        if len(line) == 2:
            unique_values.append(line[0])
            target_values.append(line[1])

# Count how often values from `unique_values` appear in `target_values`
# The dictionary is converted to a List[Tuple] for further processing
frequency_tuples = {
    value: target_values.count(value) for value in unique_values
}.items()

# Create a list of weighted counts
weighted_counts = []
for value, count in frequency_tuples:
    # Skip entries with a count of zero
    if count != 0:
        weighted_counts.append(int(value) * count)
del frequency_tuples

# Calculate the total sum of weighted counts
total_sum = sum(weighted_counts)

# Output the result
print(total_sum)
